### PR TITLE
Meta-4227: Generate qualifiedName for ReadMe entity at server side

### PR DESCRIPTION
**Pull request by hitarthtalavia**
There was ambiguity when ReadMe to asset were generated by user via API using their own qualifiedName pattern.

In order to have constituency in qualifiedName is generated at server side.

Linear: https://linear.app/atlanproduct/issue/META-4227

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        AtlasPerfMetrics.MetricRecorder metricRecorder = requestContext.startMetricRecord("processCreateReadme");
+        String entityQualifiedName = createQualifiedName(entity);
+
+        entity.setAttribute(QUALIFIED_NAME, entityQualifiedName);
+
+        AtlasEntityType entityType = typeRegistry.getEntityTypeByName(entity.getTypeName());
+        AtlasVertex vertex = AtlasGraphUtilsV2.findByUniqueAttributes(graph, entityType, entity.getAttributes());
+        if(vertex != null){
+            if(vertex.getProperty(STATE_PROPERTY_KEY, String.class).equals(DELETED.name())){
+                GraphTransactionInterceptor.addToVertexStateCache(vertex.getId(), AtlasEntity.Status.ACTIVE);
```
Comment (suraj5077) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:73]: Why is this needed? Can't find reading from this cache anywhere?

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                restoreHandlerV1.restoreEntities(Collections.singletonList(vertex));
+            }
+            String localGuid = context.getDiscoveryContext().getReferencedGuids().get(((AtlasObjectId) entity.getRelationshipAttribute(ASSET)).getGuid());
```
Comment (suraj5077) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:76]: Instead of localGuid we can have a more meaningful name here.

Comment (suraj5077) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:76]: Will break here incase getRelationshipAttribute returns null, have we tested for cases where Readme asset is created via API without any relationshipAttribute? If it is possible, we need to handle for such cases.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+        entity.setAttribute(QUALIFIED_NAME, vertexQnName);
+        requestContext.recordEntityUpdate(entityRetriever.toAtlasEntityHeader(vertex, entity.getAttributes().keySet()));
+        requestContext.cacheDifferentialEntity(entity);
+        requestContext.endMetricRecord(metricRecorder);
+    }
+
+    public String createQualifiedName(AtlasEntity entity) throws AtlasBaseException {
+        AtlasObjectId asset = (AtlasObjectId) entity.getRelationshipAttribute(ASSET);
+        String guid = asset.getGuid();
```
Comment (suraj5077) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:97]: Can break here if relationshipAttribute for Readme is not present.

Comment (hitarthtalavia) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:99]: check for asset is added in `processCreateReadme`

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java (hunk 1 above)
Comment (hitarthtalavia) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java]: removed the use of cache and directly updating the status of vertex.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java (hunk 2 above)
Comment (hitarthtalavia) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java]: localGuid -> internalGuidOfReadme

Comment (hitarthtalavia) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java]: asset check done

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            AtlasPerfMetrics.MetricRecorder metricRecorder = requestContext.startMetricRecord("processCreateReadme");
+            String entityQualifiedName = createQualifiedName(asset);
+
+            entity.setAttribute(QUALIFIED_NAME, entityQualifiedName);
+
+            AtlasEntityType entityType = typeRegistry.getEntityTypeByName(entity.getTypeName());
+            AtlasVertex vertex = AtlasGraphUtilsV2.findByUniqueAttributes(graph, entityType, entity.getAttributes());
+            if (vertex != null) {
+                if (vertex.getProperty(STATE_PROPERTY_KEY, String.class).equals(DELETED.name())) {
+                    vertex.setProperty(STATE_PROPERTY_KEY, ACTIVE.name());
```
Comment (suraj5077) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:75]: RestoreHandler will update status to Active, this line not needed

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                    restoreHandlerV1.restoreEntities(Collections.singletonList(vertex));
+                }
+                String internalGuidOfReadme = context.getDiscoveryContext().getReferencedGuids().get(asset.getGuid());
```
[the remark this answers is not in the document]
Comment (hitarthtalavia) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:78]: Here, I'm getting readme's local guid. 
After making post request for readme, response have `guidAssignements` for readme which shows reference of `localGuid` to the `readme's guid`. Since we're updating the existing vertex so guid should remain same but reference in guidAssignments are newly created (`localGuid` -> `new guid of readme`) So to update guidAssignments I'm taking localGuid.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java (hunk 4 above)
Comment (hitarthtalavia) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:75]: Removed this line and added `GraphTransactionInterceptor.addToVertexStateCache(vertex.getId(), AtlasEntity.Status.ACTIVE);`
cache will store the state of vertex, this will give us the correct response when API request is made.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                break;
+            case UPDATE:
+                processUpdateReadme(entity, context.getVertex(entity.getGuid()));
+                break;
+        }
+    }
+
+    private void processCreateReadme(AtlasEntity entity, EntityMutationContext context) throws AtlasBaseException {
+        AtlasObjectId asset = (AtlasObjectId) entity.getRelationshipAttribute(ASSET);
+        if (asset != null) {
```
[the remark this answers is not in the document]
Comment (hitarthtalavia) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/readme/ReadmePreProcessor.java:65]: Readme won't be created if no asset is mentioned.

**Review by suraj5077 — APPROVED** [review]

**Comment by suraj5077** [comment]
LGTM
